packages/tpetra/core/example/advanced/Benchmarks/CrsMatrixMatVec.cpp

```diff
@@ -0,0 +1,380 @@
+  Teuchos::OSTab tab1 (out);
+  out << "numTrials: " << opts.numTrials << endl
+      << "lclNumRows: " << opts.lclNumRows << endl
+      << "numEntPerRow: " << opts.numEntPerRow << endl
+      << endl;
+}
+
+// Return a pointer (RCP is like std::shared_ptr) to an output stream.
+// It prints on Process 0 of the given MPI communicator, but ignores
+// all output on other MPI processes.
+Teuchos::RCP<Teuchos::FancyOStream>
+getOutputStream (const Teuchos::Comm<int>& comm)
+{
+  using Teuchos::getFancyOStream;
+
+  const int myRank = comm.getRank ();
+  if (myRank == 0) {
+    // Process 0 of the given communicator prints to std::cout.
+    return getFancyOStream (Teuchos::rcpFromRef (std::cout));
+  }
+  else {
+    // A "black hole output stream" ignores all output directed to it.
+    return getFancyOStream (Teuchos::rcp (new Teuchos::oblackholestream ()));
+  }
+}
+
+// Get a Tpetra::CrsGraph for use in benchmarks.  This method takes
+// parameters that come from the command-line options read in by
+// parseCmdLineOpts.
+Teuchos::RCP<Tpetra::CrsGraph<> >
+getTpetraGraph (const Teuchos::RCP<const Teuchos::Comm<int> >& comm,
+                const CmdLineOpts& opts)
+{
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  typedef Tpetra::Map<> map_type;
+  typedef Tpetra::CrsGraph<> graph_type;
+  typedef Tpetra::Map<>::local_ordinal_type LO;
+  typedef Tpetra::Map<>::global_ordinal_type GO;
+  typedef Tpetra::global_size_t GST;
+
+  const LO lclNumRows = opts.lclNumRows;
+  const GST gblNumRows = static_cast<GST> (opts.lclNumRows) *
+    static_cast<GST> (comm->getSize ());
+  const GO indexBase = 0;
+
+  // A Map describes a distribution of data over MPI processes.
+  // This "row Map" will describe the distribution of rows of the
+  // sparse graph that we will create.
+  RCP<const map_type> rowMap =
+    rcp (new map_type (gblNumRows, static_cast<size_t> (lclNumRows),
+                       indexBase, comm));
+  const GO gblNumCols = static_cast<GO> (rowMap->getGlobalNumElements ());
+  // Create the graph structure of the sparse matrix.
+  RCP<graph_type> G =
+    rcp (new graph_type (rowMap, opts.numEntPerRow,
+                         Tpetra::StaticProfile));
+  // Fill in the sparse graph.
+  Teuchos::Array<GO> gblColInds (opts.numEntPerRow);
+  for (LO lclRow = 0; lclRow < lclNumRows; ++lclRow) { // for each of my rows
+    const GO gblInd = rowMap->getGlobalElement (lclRow);
+    // Just put some entries in the graph.  The actual column
+    // indices don't matter so much, as long as they make the
+    // resulting matrix square and don't go out of bounds.
+    for (LO k = 0; k < static_cast<LO> (opts.numEntPerRow); ++k) {
+      const GO curColInd = (gblInd + static_cast<GO> (3*k)) % gblNumCols;
+      gblColInds[k] = curColInd;
+    }
+    G->insertGlobalIndices (gblInd, gblColInds ());
+  }
+  // Make the graph ready for use by CrsMatrix.
+  G->fillComplete ();
+  return G;
+}
+
+// Get a Tpetra::BlockCrsMatrix for use in benchmarks.
```

Comment:
It's a CrsMatrix, not a BlockCrsMatrix.  Also, we can return CrsMatrix by value now; it behaves somewhat like a Kokkos::View.

Reply:
I searched for "block" in the file and removed a bunch of other misleading occurrences.